File: 890-find-and-replace-pattern/890-find-and-replace-pattern.py
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        n = len(pattern)
        
        ans = []
        for word in words:
            pchar_map, wchar_map = {}, {}
            for i in range(n):
                if word[i] in wchar_map and wchar_map[word[i]] != pattern[i]:
                    break
                
                if pattern[i] in pchar_map and pchar_map[pattern[i]] != word[i]:
                    break
                    
                wchar_map[word[i]] = pattern[i]
                pchar_map[pattern[i]] = word[i]
            else:
                ans.append(word)
        return ans

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.java
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> ans = new ArrayList<>();
        for(String word : words){
            Map<Character, Character> pCharMap = new HashMap<>();
            Map<Character, Character> wCharMap = new HashMap<>();
            int i = 0;
            for(; i < pattern.length(); i++){
                if(pCharMap.getOrDefault(pattern.charAt(i), word.charAt(i)) != word.charAt(i)){
                    break;
                }
                
                if(wCharMap.getOrDefault(word.charAt(i), pattern.charAt(i)) != pattern.charAt(i)){
                    break;
                }
                
                pCharMap.put(pattern.charAt(i), word.charAt(i));
                wCharMap.put(word.charAt(i), pattern.charAt(i));
            }
            
            if(i == pattern.length()){
                ans.add(word);
            }
        }
        return ans;
    }
}